### DISCUSSION TITLE
Is there a specific reason for the possible aspect ratios used in bucketing?

### DISCUSSION THREAD

**Discussion by minienglish1** [d8ahazard/sd_dreambooth_extension#987] [Q&A] [answered]
Not a complaint or a request.  Just trying to learn.

Is there a specific reason for the possible aspect ratios used in bucketing?

The code has: aspect_ratios = [(16, 9), (5, 4), (4, 3), (3, 2), (2, 1), (1, 1)]
With the requirement that the short dimension be divisible by 64, there should be lots more possible aspect ratios.

Since I'm training at 1024, to get potentially, better image cropping via bucketing, I added more aspect ratios:
aspect_ratios = [(16, 15), (8, 7), (16, 13), (16, 11), (8, 5), (16, 7), (8, 3), (16, 5), (4, 1), (16, 9), (4, 3), (3, 2), (2, 1), (1, 1)]

Nothing seemed to break, but my images only additionally used the (1, 4) aspect ratio.
Maybe none of my images fit the other aspect ratios or maybe there is something I don't understand about the code that limits possible aspect ratios.

**Comment by d8ahazard** [comment, marked as answer]
Mostly just to help with increasing the possible number of images per
bucket. The logic was that most "real-world" images should fall closely
under one of the provided aspect ratios, especially at smaller resolutions
where the 64 pixel step became closer to whatever we were using as inputs
anyway. And, if you use larger input images, then the minor difference in
pixels between images cropped in various programs becomes more a hindrance
than a benefit.

At least, that was my train of thought when I changed it. I'm open to
discussing a Pull Request to add "more" aspect ratios if it helps things,
but I think as you demonstrated, at some point, it doesn't make a huge
difference. ;)

**Reply by cerega66** [reply]
[d8ahazard](https://github.com/d8ahazard) is there any compensation algorithm for images of different resolutions? After all, images of different resolutions require a different number of steps (epochs) of training due to different sizes of matrices. If I train 1024 and I get 2 images, one 1024x1024 and the second 1024x512, then for the first I need to go through 2 times more epochs than for the second. And if there are too many such moments, it can lead to the fact that some images will be overtrained and others undertrained.

**Reply by minienglish1** [reply]
You're right, I don't think adding a large amount of resolutions would help.  Possibly add (1, 4), just to help people coming from other trainers.  The only reason I even have (1, 4) images in my database is because another trainer supports that aspect ratio.  Otherwise, I would never even considered it.

I'm not a programmer and I don't even know how to do a pull request.  After I finish studying ML, I'll start learning programming.

However, I am open to helping out updating the wiki, readme, and the like.  If you can point me in the right direction on how to learn to do so.  I've been learning a lot recently, so I could update it as I learn.  Would also help me to organize my notes.

**Reply by d8ahazard** [reply]
> You're right, I don't think adding a large amount of resolutions would help. Possibly add (1, 4), just to help people coming from other trainers. The only reason I even have (1, 4) images in my database is because another trainer supports that aspect ratio. Otherwise, I would never even considered it.
> 
> I'm not a programmer and I don't even know how to do a pull request. After I finish studying ML, I'll start learning programming.
> 
> However, I am open to helping out updating the wiki, readme, and the like. If you can point me in the right direction on how to learn to do so. I've been learning a lot recently, so I could update it as I learn. Would also help me to organize my notes.

To submit a PR, you'd do the following...

Fork the *dev* branch. You do that by going to the main page of the repo, picking the "dev" branch, and then clicking "fork" in the top right.

Pull your own fork to your computer...you don't necessarily have to even install it in Auto1111, but testing is *usually* recommended.

Add your modified aspect ratio to the relevant line of code where I have them defined.

Push to your forked repo.

Go to the main github page of your forked repo, the top should say "Had recent pushes", and there will be a button to create a pull request. Click that, submit to the main repo.

Vbladow...you're now an "official" contributor to the project. ;)

**Reply by minienglish1** [reply]
Thanks. Just caught a cold, so my head is out of it.  And I've got a busy weekend ahead of me at work, so it'll be lots of fun to be sick at work.  I'll try playing around with it on Tuesday or Wednesday.

**Reply by minienglish1** [reply]
> To submit a PR, you'd do the following...
> 
> Fork the _dev_ branch. You do that by going to the main page of the repo, picking the "dev" branch, and then clicking "fork" in the top right.
> 
> Pull your own fork to your computer...you don't necessarily have to even install it in Auto1111, but testing is _usually_ recommended.
> 
> Add your modified aspect ratio to the relevant line of code where I have them defined.
> 
> Push to your forked repo.
> 
> Go to the main github page of your forked repo, the top should say "Had recent pushes", and there will be a button to create a pull request. Click that, submit to the main repo.
> 
> Vbladow...you're now an "official" contributor to the project. ;)

Slowly making my way through it.  Created my fork.  Learning how to pull and push.

However, when I did my test with only adding (1, 4), the resulting bucketing was different.  So maybe some of the other aspect ratios are being used.  I'll need to check to see what is going on.   I also might as well start using the dreambooth dev branch with the webui for easier testing.

**Comment by d8ahazard** [comment]
Yes, there is. The number of iterations per bucket is tracked and balanced.
